File: sol/1367/1367.py
class Solution:
    def isSubPath(self, head: Optional[ListNode], root: Optional[TreeNode]) -> bool:
        def checker(node: Optional[TreeNode], ptr: Optional[ListNode]) -> bool:
            if ptr is None:
                return True
            if node is None or node.val != ptr.val:
                return False
            return checker(node.left, ptr.next) or checker(node.right, ptr.next)
        st = [root]
        while len(st) != 0:
            node = st.pop()
            if checker(node, head):
                return True
            if node.left is not None:
                st.append(node.left)
            if node.right is not None:
                st.append(node.right)
        return False

File: sol/1367/1367.cpp
struct Solution {
    bool checker(TreeNode *node, ListNode *ptr){
        if (ptr == nullptr){
            return true;
        }
        if (node == nullptr || node->val != ptr->val){
            return false;
        }
        return this->checker(node->left, ptr->next) || this->checker(node->right, ptr->next);
    }
    bool isSubPath(ListNode* head, TreeNode* root) {
        vector<TreeNode *> st = {root};
        while (!st.empty()){
            TreeNode *node = st.back();
            st.pop_back();
            if (this->checker(node, head)){
                return true;
            }
            if (node->left != nullptr){
                st.push_back(node->left);
            }
            if (node->right != nullptr){
                st.push_back(node->right);
            }
        }
        return false;
    }
};
